lib/autosign/validator.rb

```diff
@@ -108,103 +46,13 @@ def self.any_validator(challenge_password, certname, raw_csr)
-    end
-
-    # provide a merged settings hash of default settings for a validator,
-    # config file settings for the validator, and override settings defined in
-    # the validator.
-    #
-    # Do not override this in child classes. If you need to set
-    # custom config settings, override the get_override_settings method.
-    # The section of the config file this reads from is the same as the name
-    # method returns.
-    #
-    # @return [Hash] of config settings
-    def settings
-      @log.debug 'merging settings'
-      setting_sources = [get_override_settings, load_config, default_settings]
-      merged_settings = setting_sources.inject({}) { |merged, hash| merged.deep_merge(hash) }
-      @log.debug 'using merged settings: ' + merged_settings.to_s
-      @log.debug 'validating merged settings'
-      if validate_settings(merged_settings)
-        @log.debug 'successfully validated merged settings'
-        merged_settings
-      else
-        @log.warn 'validation of merged settings failed'
-        @log.warn "unable to validate settings in #{name} validator"
-        raise 'settings validation error'
-      end
-    end
-
-    # (optionally) override this from a child class to set config defaults.
-    # These will be overridden by config file settings.
-    #
-    # Override this when inheriting if you need to set config defaults.
-    # For example, if you want to pull settings from zookeeper, this would
-    # be a good place to do that.
-    #
-    # @return [Hash] of config settings
-    def default_settings
-      {}
-    end
-
-    # (optionally) override this to perform validation checks on the merged
-    # config hash of default settings, config file settings, and override
-    # settings.
-    # @return [True, False]
-    def validate_settings(settings)
-      settings.is_a?(Hash)
-    end
-
-    # load any required configuration from the config file.
-    # Do not override this in child classes.
-    # @return [Hash] configuration settings from the validator's section of the config file
-    def load_config
-      @log.debug 'loading validator-specific configuration'
-      config = Autosign::Config.new
-
-      if config.settings.to_hash[name].nil?
-        @log.warn 'Unable to load validator-specific configuration'
-        @log.warn "Cannot load configuration section named '#{name}'"
-        {}
-      else
-        @log.debug 'Set validator-specific settings from config file: ' + config.settings.to_hash[name].to_s
-        config.settings.to_hash[name]
-      end
-    end
-
-    # (optionally) override this from child classes to get custom configuration
-    # from a validator.
-    #
-    # This is how you override defaults and config file settings.
-    # @return [Hash] configuration settings
-    def get_override_settings
-      {}
+    def self.validator_classes
+      validators = Dir.glob(File.join(__dir__, 'validators', '*')).sort.each {|k| require k }
```

Comment:
Why is this dynamic? It would be easier to read and understand if we just enumerated them.

Reply:
At one point there was thought to allow external gems to influence this list. So if a loaded class is in the object space this method would find it and include it as part of a validator.   Currently there is no dynamic way to load external gems so the user would have to require their special validator before this code is executed.   It is a cheap solution for pluggable validators.